spread.yaml

```diff
@@ -1,40 +1,109 @@
-            echo "Discard called"
-            make -C tests/lib background-stop
-        systems:
-            - core20-amd64:
-                username: ubuntu
-                password: ubuntu
-path: /tmp/core20
+  google-nested:
+    type: google
+    key: '$(HOST: echo "$SPREAD_GOOGLE_KEY")'
+    location: snapd-spread/us-east1-b
+    plan: n1-standard-2
+    halt-timeout: 2h
+    systems:
+      - ubuntu-20.04-64:
+          image: ubuntu-2004-64-virt-enabled
+          storage: 20G
+
+  qemu-nested:
+    type: qemu
+    # TODO:UC20: uc20 needs 2G or grub will not loopback the kernel snap
+    #            we use nested VM's, so use 4G for the host until we can boot
+    #            uc20 with less RAM
+    memory: 4G
+    systems:
+      - ubuntu-20.04-64:
+          username: ubuntu
+          password: ubuntu
+
+  lxd-nested:
+    type: lxd
+    container-profiles: core20
+    systems:
+      - ubuntu-20.04:
+          image: ucspread
+
+path: /home/core20
+
 exclude:
-    - "*.raw"
-    - "*.snap"
-    - Makefile
-    - README.md
-    - hooks
-    - parts
-    - prime
-    - snap
-    - spread.yaml
-    - stage
-kill-timeout: 20m
+  - .git
+  - "*.o"
+  - "*.a"
+  - "*.snap"
+  - "*.img"
+
 prepare: |
-    true
-restore: |
-    true
+  # NOTE: This part of the code needs to be in spread.yaml as it runs before
+  # the rest of the source code (including the tests/lib directory) is
+  # around. The purpose of this code is to fix some connectivity issues setup 
+  # spread functions to be more usable outside of the explicit task.yaml 
+  # sections.
+
+  # apt update is hanging on security.ubuntu.com with IPv6, prefer IPv4 over IPv6
+  cat <<EOF > gai.conf
+  precedence  ::1/128       50
+  precedence  ::/0          40
+  precedence  2002::/16     30
+  precedence ::/96          20
+  precedence ::ffff:0:0/96 100
+  EOF
+  if ! mv gai.conf /etc/gai.conf; then
+      echo "/etc/gai.conf is not writable, ubuntu-core system? apt update won't be affected in that case"
+      rm -f gai.conf
+  fi
+
+  # Take the MATCH and REBOOT functions from spread and allow our shell
+  # scripts to use them as shell commands. The replacements are real
+  # executables in tests/lib/bin (which is on PATH) but they source
+  # spread-funcs.sh written here, base on the definitions provided by SPREAD.
+  # This ensures that 1) spread functions define the code 2) both MATCH and
+  # REBOOT are executables and not functions, and can be called from any
+  # context.
+  type MATCH | tail -n +2 > "$TESTSLIB"/spread-funcs.sh
+  unset MATCH
+  type REBOOT | tail -n +2 >> "$TESTSLIB"/spread-funcs.sh
+  unset REBOOT
+
+  # prepare common uc20 image setup by repacking snaps, etc
+  "$TESTSLIB"/prepare-uc20.sh
+restore-each: |
+  # delete the nested VM image after each task finishes so we don't use too much
+  # disk space
+  . "$TESTSLIB/nested.sh"
+  cleanup_nested_core_vm
+debug-each: |
+  . "$TESTSLIB/nested.sh"
+  print_nested_status
+
+warn-timeout: 40m
```

Comment:
40m for the warn-timeout seems to be too much compared with the kill timeout. why are you using that value?

Reply:
This value is reused from core-initrd, and the reason is that running qemu-nested can take quite a while - I've experienced it taking up to 33m so I believe this value should be fine.